draft/XXXX-robust-api/README.md

```diff
@@ -0,0 +1,146 @@
+---
+RFC: XXXX
+Author: Jacob Nesbitt
+Created: 2020-06-29
+Created: 2020-12-12
+Last-Modified: 2020-12-12
+---
+
+# RFC XXXX: RFC Template
+
+This RFC outlines a robust way to handle table/graph creation and upload, allowing for arbitrarily large datasets to be uploaded in any of the supported formats.
+
+## Motivation
+
+As the size of our datasets has grown, our limited approach to handling data upload and table/graph creation has proved to be inefficient. The need has arose for an API that can handle larger data, with more operations. The proposals outlined below will make that possible.
+
+# Proposal
+
+## Upload
+Currently, our upload system deals with upload, data processing, and table creation, all at once and in the same request. Since we deploy on heroku, and we need to keep our requests under 30 seconds (as we should regardless), this severely limits both the size of the data we can upload. In order to address this, we can make use of Amazon S3. The new upload process would follow this rough path:
+
+
+### Create an upload
+
+`POST /api/workspaces/{workspace}/upload/create`
+
+Create an upload, returning an `Upload` object that (among other things) will contain a list of pre-signed URLs, that are to be used for uploading each part of the file to S3. **Note:** This means that the amount of parts / size of each part is determined by the server. The parameters for this call will include the following
+
+* `dataType` - The data type of the file to upload, which must be one of the supported (or installed, if using a plugin system) data types. E.g. `csv`, `d3_json`, `nested_json`, `newick`, etc. If this parameter does not match one of the supported data types, a `4XX` error is thrown.
```

Comment:
I wonder if this should be passed on *finalize*, rather than upload creation, *unless* we want to validate the data before it's uploaded (which seems to be almost a necessity for very large files).

I also wonder if we should create offline tools for validating preflight data... It would be easier to write and run that sort of validator as, e.g., a Python script after you've prepped your data but before you even bring it near Multinet. This notion of "preflight validation" is something we should discuss.

Reply:
As far as passing it on finalize vs on upload creation, I don't see a particular reason to pass it on finalize, since you'll know the data type at upload selection time. IMO, if we know this information we should pass it onto the API immediately, rather than wait.

I think the thoughts on offline tools are tangent to this issue. When it comes to our API, we shouldn't allow invalid data (from any source) to be inserted into our system. The only real way to do this is to validate the data in the API itself.

Reply:
> As far as passing it on finalize vs on upload creation, I don't see a particular reason to pass it on finalize, since you'll know the data type at upload selection time.

The only reason would be to prevent accidents where you specified the wrong format at that early stage. For instance, if you upload a massive CSV file, but you mistakenly tell the API that it's a Newick file or something like that.

Basically, my thought process is that uploading a file to S3 can be done without knowing what type of file it is, so to require the minimum amount of information to do so would mean deferring the file format info to a later time.

> IMO, if we know this information we should pass it onto the API immediately, rather than wait.

But why? The server has to cache this information to make use of it at the right time. Furthermore, if you know the information now, then you'll know it later (when it's needed) as well.

I don't think this decision should hinder the initial implementation, so go with what you think is best (but consider my arguments in case they're persuasive). If we do it one way and decide it's done better another way, we can change it later. The discussion is valuable though. IMO, the fact that we're debating when to specify the file type points to a weakness in our overall design: I more and more am doubtful that dealing with file formats is something the server should be doing at all, but that's a totally separate topic of discussion.

> 
> I think the thoughts on offline tools are tangent to this issue. When it comes to our API, we shouldn't allow invalid data (from any source) to be inserted into our system. The only real way to do this is to validate the data in the API itself.

Agreed. My tangent thinking, as you put it, was musing on whether we can prevent disasters (i.e., having to re-upload massive files with minor mistakes corrected) by providing extra, preclient tooling. On that view, the API-side validation is to gate bad data from getting in, while the preflight validation would be to save the client frustration. I will file an issue around that and we can drop discussion of it for now.

Reply:
I agree, we should discuss this in a separate issue.

Reply:
See #26.